src/fundus/scraping/pipeline.py

```diff
@@ -1,47 +1,163 @@
-from typing import Iterator, List, Literal, Optional, Set, Tuple, Type, Union
+import asyncio
+import time
+from typing import (
+    AsyncIterator,
+    Iterable,
+    Iterator,
+    List,
+    Literal,
+    Optional,
+    Protocol,
+    Set,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
 
 import more_itertools
 
+from fundus.logging import basic_logger
 from fundus.publishers.base_objects import PublisherEnum
 from fundus.scraping.article import Article
-from fundus.scraping.filter import ExtractionFilter
-from fundus.scraping.html import URLSource
+from fundus.scraping.filter import ExtractionFilter, URLFilter
+from fundus.scraping.html import URLSource, session_handler
 from fundus.scraping.scraper import Scraper
+from fundus.utils.more_async import async_next, batched_async_interleave
 from fundus.utils.validation import listify
 
 
+class Delay(Protocol):
+    """Protocol to define crawl delays between batches."""
+
+    def __call__(self) -> float:
+        """Yields back a float to specify crawler delay for current batch in seconds
+
+        This inludes the execuion time between batches.
+        The crawler will wait max(execution_time, delay)
+
+        Examples:
+            >>> import random
+            >>> delay: Delay = lambda: random.randint(1, 1000)/1000.
+            This will wait a random amount of milliseconds between batches.
+
+        Returns:
+
+        """
+        ...
+
+
 class Pipeline:
     def __init__(self, *scrapers: Scraper):
+        """Basic pipeline to utilize scrapers.
+
+        Because scrapers are implemented asynchronous this pipeline handles
+        the necessary event loops and program logic to download articles
+        in batches asynchronously.
+
+        Args:
+            *scrapers (Scraper): The scrapers which should be used.
+        """
         self.scrapers: Tuple[Scraper, ...] = scrapers
 
     def run(
         self,
         error_handling: Literal["suppress", "catch", "raise"],
         max_articles: Optional[int] = None,
         extraction_filter: Optional[ExtractionFilter] = None,
-        batch_size: int = 10,
+        delay: Optional[Delay] = lambda: 0.1,
+        url_filter: Optional[URLFilter] = None,
+        only_unique: bool = True,
     ) -> Iterator[Article]:
-        scrape_map = map(
-            lambda x: x.scrape(
-                error_handling=error_handling, batch_size=batch_size, extraction_filter=extraction_filter
-            ),
-            self.scrapers,
-        )
-        robin = more_itertools.interleave_longest(*tuple(scrape_map))
-
-        if max_articles:
-            while max_articles:
-                try:
-                    yield next(robin)
-                except StopIteration:
-                    pass
-                max_articles -= 1
+        """Yields articles from initialized scrapers.
```

Comment:
I have not looked at the remaining docstrings yet. But a check with Grammarly would surely help.

Reply:
I agree.